akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala

```diff
@@ -1635,6 +1637,21 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
     new Flow(delegate.zipWithMat[Out2, Out3, M, M2](that)(combinerToScala(combine))(combinerToScala(matF)))
 
   /**
+   * Combine the elements of current flow into a stream of tuples consisting
+   * of all elements paired with their index. Indices start at 0.
+   *
+   * '''Emits when''' upstream emits an element and is paired with their index
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def zipWithIndex: Flow[In, (Out, Long), Mat] =
+    new Flow(delegate.zipWithIndex)
```

Comment:
Doesn't need to be in this PR but wouldn't a variation corresponding to `zipWith` be nice? I mean something like `zipWithIndex(combine: (In, Long) => Out)` so that it would be easy to use the index with a custom data structure without going via tuple? Would make the Java API easier to create as well (see yet to be written comment below).